File: workers/go/internal/worker/emit.go
package worker

import (
	"encoding/json"
	"fmt"
	"io"
)

type eventEmitter struct {
	encoder *json.Encoder
	scanID  string
	seq     uint64
}

func Emit(writer io.Writer, scanID string, result Result) error {
	emitter := newEventEmitter(writer, scanID)
	if err := emitter.write("scan_started", map[string]any{
		"root": result.Root, "profile_ids": []string{result.Profile.ID}, "project_code_executed": false, "safe_mode": true,
	}); err != nil {
		return err
	}
	if err := emitter.write("profile_declared", map[string]any{"profile": result.Profile}); err != nil {
		return err
	}
	for _, node := range result.Nodes {
		if err := emitter.write("node_upsert", map[string]any{"node": node}); err != nil {
			return err
		}
	}
	for _, site := range result.Sites {
		if err := emitter.write("dependency_site", map[string]any{"site": site}); err != nil {
			return err
		}
	}
	for _, edge := range result.Edges {
		if err := emitter.write("edge_upsert", map[string]any{"edge": edge}); err != nil {
			return err
		}
	}
	for _, diagnostic := range result.Diagnostics {
		if err := emitter.write("diagnostic", map[string]any{"diagnostic": diagnostic}); err != nil {
			return err
		}
	}
	for _, file := range result.Files {
		fields := map[string]any{
			"path": file.Path, "discovered_sites": file.DiscoveredSites, "emitted_sites": file.EmittedSites,
			"skipped_sites": file.SkippedSites, "skipped": file.Skipped,
		}
		if file.Reason != "" {
			fields["reason"] = file.Reason
		}
		if err := emitter.write("file_completed", fields); err != nil {
			return err
		}
	}
	if err := emitter.write("profile_completed", map[string]any{"profile_id": result.Profile.ID, "coverage": result.Coverage}); err != nil {
		return err
	}
	return emitter.write("scan_completed", map[string]any{"coverage": result.Coverage})
}

func EmitFailure(writer io.Writer, scanID, root string, scanErr error) error {
	emitter := newEventEmitter(writer, scanID)
	if err := emitter.write("scan_started", map[string]any{"root": root, "profile_ids": []string{}, "project_code_executed": false, "safe_mode": true}); err != nil {
		return err
	}
	diagnostic := Diagnostic{
		ID: stableID("diagnostic", "scan-failure", root, scanErr.Error()), Code: "scan_failure",
		Severity: "error", Message: scanErr.Error(), Recoverable: false,
	}
	if err := emitter.write("diagnostic", map[string]any{"diagnostic": diagnostic}); err != nil {
		return err
	}
	coverage := Coverage{ProjectCodeExecuted: false, Completeness: []string{}, Reasons: []string{"scan-failure"}}
	return emitter.write("scan_completed", map[string]any{"coverage": coverage})
}

func newEventEmitter(writer io.Writer, scanID string) *eventEmitter {
	encoder := json.NewEncoder(writer)
	encoder.SetEscapeHTML(false)
	return &eventEmitter{encoder: encoder, scanID: scanID}
}

func (e *eventEmitter) write(event string, fields map[string]any) error {
	e.seq++
	record := map[string]any{
		"event": event, "protocol_version": ProtocolVersion, "scan_id": e.scanID,
		"adapter": AdapterName, "adapter_version": AdapterVersion, "seq": e.seq,
	}
	for key, value := range fields {
		record[key] = value
	}
	if err := e.encoder.Encode(record); err != nil {
		return fmt.Errorf("encode %s event: %w", event, err)
	}
	return nil
}

File: workers/rust/src/emit.rs
use crate::{ADAPTER, ADAPTER_VERSION, ScanResult};
use anyhow::{Context, Result};
use depgraph_protocol::{
    CommonFields, DependencySiteEvent, DiagnosticEvent, EdgeUpsert, FileCompleted, NodeUpsert,
    ProfileCompleted, ProfileDeclared, ProtocolEvent, ProtocolValidator, ScanCompleted,
    ScanStarted,
};

/// Build a complete, typed, self-validated v1.0 event stream.
pub fn build_events(scan_id: &str, result: &ScanResult) -> Result<Vec<ProtocolEvent>> {
    let mut seq = 0_u64;
    let mut common = || {
        seq += 1;
        CommonFields {
            protocol_version: depgraph_protocol::PROTOCOL_VERSION.into(),
            scan_id: scan_id.into(),
            adapter: ADAPTER.into(),
            adapter_version: ADAPTER_VERSION.into(),
            seq,
        }
    };

    let mut events = vec![
        ProtocolEvent::ScanStarted(ScanStarted {
            common: common(),
            root: result.root.clone(),
            project_code_executed: false,
            safe_mode: true,
        }),
        ProtocolEvent::ProfileDeclared(ProfileDeclared {
            common: common(),
            profile: result.profile.clone(),
        }),
    ];
    events.extend(result.nodes.iter().cloned().map(|node| {
        ProtocolEvent::NodeUpsert(NodeUpsert {
            common: common(),
            node,
        })
    }));
    events.extend(result.sites.iter().cloned().map(|site| {
        ProtocolEvent::DependencySite(DependencySiteEvent {
            common: common(),
            site,
        })
    }));
    events.extend(result.edges.iter().cloned().map(|edge| {
        ProtocolEvent::EdgeUpsert(EdgeUpsert {
            common: common(),
            edge,
        })
    }));
    events.extend(result.diagnostics.iter().cloned().map(|diagnostic| {
        ProtocolEvent::Diagnostic(DiagnosticEvent {
            common: common(),
            diagnostic,
        })
    }));
    events.extend(result.files.iter().map(|file| {
        ProtocolEvent::FileCompleted(FileCompleted {
            common: common(),
            path: file.path.clone(),
            discovered_sites: file.discovered_sites,
            emitted_sites: file.emitted_sites,
            skipped_sites: file.discovered_sites.saturating_sub(file.emitted_sites),
            skipped: file.skipped,
            reason: file.reason.clone(),
        })
    }));
    events.push(ProtocolEvent::ProfileCompleted(ProfileCompleted {
        common: common(),
        profile_id: result.profile.id.clone(),
        coverage: result.coverage.clone(),
    }));
    events.push(ProtocolEvent::ScanCompleted(ScanCompleted {
        common: common(),
        coverage: result.coverage.clone(),
    }));

    let mut validator = ProtocolValidator::new();
    for event in events.iter().cloned() {
        validator
            .push(event)
            .context("Rust worker generated an invalid protocol event")?;
    }
    validator
        .finish()
        .context("Rust worker generated an incomplete protocol stream")?;
    Ok(events)
}
